### PR TITLE
Start using .travis.yml as a configure doctr.

### PR DESCRIPTION
**Pull request by Carreau**
My guess is it will be easier with a subparser. 

Just playing with it for now, see what's needed.

### CODE REVIEW

File: doctr/__main__.py [hunk 1]
```diff
@@ -102,6 +107,16 @@ def get_parser():
 
     return parser
 
+def get_config():
+    """
+    This load some configuration from the `travis.yml`, file is present,
+    `doctr` key if present.
```
Comment (asmeurer) [inline doctr/__main__.py]: Two backquotes for RST.

File: doctr/__main__.py [hunk 2]
```diff
@@ -102,6 +107,16 @@ def get_parser():
 
     return parser
 
+def get_config():
+    """
+    This load some configuration from the `travis.yml`, file is present,
```
Comment (asmeurer) [inline doctr/__main__.py]: .travis.yml

Comment (asmeurer) [inline doctr/__main__.py]: *if

**Comment by Carreau** [comment]
Reviving this as part of the docathon. 

I've cleaned it up a bit and will appreciate review.

**Comment by Carreau** [comment]
Oh, of course test are failing as the PR is from a branch not on this repo.

File: .travis.yml [hunk 3]
```diff
@@ -31,10 +31,14 @@ script:
       cd docs;
       make html;
       cd ..;
-      python -m doctr deploy --gh-pages-docs . --key-path deploy_key.enc;
-      python -m doctr deploy --gh-pages-docs docs --key-path deploy_key.enc;
-      python -m doctr deploy --no-require-master --gh-pages-docs "docs-$TRAVIS_BRANCH" --built-docs docs/_build/html --key-path deploy_key.enc;
-      python -m doctr deploy --no-require-master --key-path deploy_key.enc --no-sync --command "echo test";
+      if [[ -z "${DOCTR_DEPLOY_ENCRYPTION_KEY}" ]]; then
+          echo "DOCTR_DEPLOY_ENCRYPTION_KEY not set, likely building a PR from untrusted repository";
```
Comment (asmeurer) [inline .travis.yml]: We ought to be testing this in doctr itself. Is this happening to you here?

File: docs/changelog.rst [hunk 4]
```diff
@@ -2,6 +2,14 @@
  Doctr Changelog
 =================
 
+Master
+======
+
+- ``setup_GitHub_push`` now takes a ``branch_whitelist`` parameter instead of
+  of a ``require_master``
+- ``.travis.yml`` can be used to store some of doctr configuration in addition
+  to the command line flags. Write doctr configuration under the ``doctr`` key.
```
Comment (asmeurer) [inline docs/changelog.rst]: Add <code>:issue:\`137\`</code> to these.

File: docs/commandline.rst [hunk 5]
```diff
@@ -4,3 +4,37 @@
...
+-------------
+
+In addition to command line arguments you can configure ``doctr`` using the
+``.travis.yml`` files. Command line arguments take precedence over the value
+present in the configuration file.
+
+The configuration parameters available from the ``.travis.yml`` file often
+mirror their command line siblings except dashes ``-`` are replaced by
+underscores ``_``. Some parameters are not available from configuration files
+for security reasons.
```
Comment (asmeurer) [inline docs/commandline.rst]: Can we be more specific here? Are there any values that aren't the same after replacing `-` with `_`? Which flags aren't available for security reasons?

File: docs/commandline.rst [hunk 6]
```diff
@@ -4,3 +4,37 @@
...
+
+Use a ``doctr`` section in your ``travis.yml`` file to store your doctr
+configuration:
+
+.. code::
```
Comment (asmeurer) [inline docs/commandline.rst]: You can set this as YAML language (though IIRC the highlighting might not actually work).

File: docs/commandline.rst [hunk 7]
```diff
@@ -4,3 +4,37 @@
...
+
+  - language: python
+  - script: py.test
+  - doctr:
+    gh_pages_docs: 'documentation'
+    key_path : '/absolute/key/path.key'
```
Comment (asmeurer) [inline docs/commandline.rst]: Absolute key path? The key path should be relative to the github repo.

File: docs/commandline.rst [hunk 8]
```diff
@@ -4,3 +4,37 @@
...
+
+Use a ``doctr`` section in your ``travis.yml`` file to store your doctr
+configuration:
+
+.. code::
+
+  - language: python
+  - script: py.test
+  - doctr:
+    gh_pages_docs: 'documentation'
```
Comment (asmeurer) [inline docs/commandline.rst]: This will change with https://github.com/drdoctr/doctr/pull/157 (which will be merged soon).

File: .travis.yml (hunk 3 above)
Comment (Carreau) [inline .travis.yml]: Hum, yes on my fork. Let me see if it was not rebased on the lastest.

File: docs/commandline.rst [hunk 9]
```diff
@@ -4,3 +4,37 @@
...
+underscores ``_``. Some parameters are not available from configuration files
+for security reasons.
+
+Use a ``doctr`` section in your ``travis.yml`` file to store your doctr
+configuration:
+
+.. code::
+
+  - language: python
+  - script: py.test
```
Comment (asmeurer) [inline docs/commandline.rst]: You still have to actually call `doctr deploy` in the script though, right?

File: doctr/__main__.py [hunk 10]
```diff
@@ -102,6 +125,20 @@ def get_parser():
...
+    """
+    This load some configuration from the ``.travis.yml``, if file is present,
+    ``doctr`` key if present.
+    """
+    with open('.travis.yml') as travis_config_file:
+        travis_config = yaml.safe_load(travis_config_file.read())
+
+    config = travis_config.get('doctr',{})
+
+    if type(config) == list:
```
Comment (asmeurer) [inline doctr/__main__.py]: `if not isinstance(config, dict)`?

File: doctr/__main__.py [hunk 11]
```diff
@@ -102,6 +125,20 @@ def get_parser():
 
     return parser
 
+def get_config():
+    """
+    This load some configuration from the ``.travis.yml``, if file is present,
+    ``doctr`` key if present.
+    """
+    with open('.travis.yml') as travis_config_file:
```
Comment (asmeurer) [inline doctr/__main__.py]: Add a check if the file exists.

File: doctr/__main__.py [hunk 12]
```diff
@@ -147,7 +191,7 @@ def deploy(args, parser):
             added, removed = [], []
 
         if args.command:
-            run(shlex.split(args.command))
+            run([os.path.expanduser(x) for x in shlex.split(args.command)])
```
Comment (asmeurer) [inline doctr/__main__.py]: Why is this change here? I don't think we should do this.

File: doctr/__main__.py [hunk 13]
```diff
@@ -102,6 +125,20 @@ def get_parser():
...
+    ``doctr`` key if present.
+    """
+    with open('.travis.yml') as travis_config_file:
+        travis_config = yaml.safe_load(travis_config_file.read())
+
+    config = travis_config.get('doctr',{})
+
+    if type(config) == list:
+        raise ValueError('config is not a dict: {}'.format(config))
+    return travis_config.get('doctr',{})
```
Comment (asmeurer) [inline doctr/__main__.py]: `return config`?

File: .travis.yml (hunk 3 above)
Comment (asmeurer) [inline .travis.yml]: "We ought" meaning we should. I don't know if we actually are (but if we aren't, we need to be). Someone actually reported this same error on SymPy, so I think there may be a real bug here.

In any case, let's fix this in a separate PR.

Comment (gforsyth) [inline .travis.yml]: We do it when we decrypt the key on Travis: https://github.com/drdoctr/doctr/blob/master/doctr/travis.py#L51

Comment (asmeurer) [inline .travis.yml]: That's only supposed to be called with `canpush` is True. https://github.com/drdoctr/doctr/blob/84d09b19b4f045c93f06516dae78c55662b3db26/doctr/travis.py#L185. So either we have a bug or Travis does (maybe `$TRAVIS_PULL_REQUEST` is not always set properly). 

It would be nice if we could keep this as an error, because it is a legitimate error if someone didn't set up their .travis.yml correctly. Maybe there is a more robust way to detect pull requests than looking at the environment variables. 

Let's open an issue for this.

File: docs/commandline.rst (hunk 5 above)
Comment (Carreau) [inline docs/commandline.rst]: > Can we be more specific here? Are there any values that aren't the same after replacing - with _? Which flags aren't available for security reasons?

Hum, I don't remember. I'm using `safe_load` from Yaml so itm might be due to that. I'll rephrase.

Comment (asmeurer) [inline docs/commandline.rst]: All the parameters are paths (strings). Unexpected stuff can happen in YAML if you don't quote a string and it looks like another literal (like a float), but that's unlikely to happen here. I can't think of any other issues.

Comment (Carreau) [inline docs/commandline.rst]: Oh, yeah. `--token`, if you allow a token option, it means pasting the token in clear in the file.

Comment (asmeurer) [inline docs/commandline.rst]: `--token` is a boolean flag. The token can only be accessed from a encrypted environment variable.

**Comment by Carreau** [comment]
Ok, I'm tired today, I did a lot of mistakes, but that should be all fixed.

**Comment by asmeurer** [comment]
I pushed the branch up to this repo so the tests can run from that.

**Comment by asmeurer** [comment]
We should probably add some configuration to our .travis.yml to test this (the existing tests will still work as they are as the command line flags override the .travis.yml keys).

**Comment by Carreau** [comment]
> We should probably add some configuration to our .travis.yml to test this (the existing tests will still work as they are as the command line flags override the .travis.yml keys).

Done, I had to add opposites of some command line flags like `--sync`/`--no-sync` to be able to overwrite config file.

**Comment by asmeurer** [comment]
Cool. By the way, it's easier to review your changes if you do them as new commits.

**Comment by Carreau** [comment]
> Cool. By the way, it's easier to review your changes if you do them as new commits.

Conflict addressed and push as multiple commits.

**Comment by asmeurer** [comment]
Let's give it a try. Also self reminder we need to add pyyaml to the conda-forge recipe when we release.
